firedrake/cofunction.py

```diff
@@ -79,6 +79,21 @@ def __del__(self):
         if hasattr(self, "_comm"):
             mpi.decref(self._comm)
 
+    def __call__(self, expr):
+        r"""Return the symbolic duality pairing, this is to evaluate this
+        :class:`.Cofunction` at an expression in the primal space, equivalent to
+        taking the dot product of the vectors of coefficients of the dual and
+        primal arguments.
+
+        :arg expr: a UFL expression to be paired with this :class:`.Cofunction`.
+
+        :returns: the symbolic duality pairing.
+        """
+        try:
+            return ufl.action(self, expr)
```

Comment:
I don't think that the assumption that a `TypeError` would be a consequence of a wrong function space is safe. Think about the case when one passes in the wrong type for `expr`.